### PR TITLE
RELATED: RAIL-2760 Move scheduled emails from gdc-dashboards to gooddata-ui-sdk

### PR DESCRIPTION
**Pull request by kandl**
- Move scheduled emails components, translations & css from gdc-dashboards to gooddata-ui-sdk
- Add necessary feature flags to sdk-backend-spi
- Include ScheduledMailDialog in DashboardView (controlled via isScheduledMailDialogVisible & relevant callback props)
- Limit supported dashboard filters (IMeasureValueFilter & IRankingFilter are not supported)
- Propagate filters from DashboardView to stored filter context in scheduled e-mail
- Add applyFiltersToScheduledMail prop, so user is able to suppress storing filters in the scheduled e-mails
- Minor bug fixes

JIRA: RAIL-2760, subtask: RAIL-2830

---

Supported PR commands:

| Command         | Description            |
| --------------- | ---------------------- |
| `ok to test`    | Re-run standard checks |
| `extended test` | BackstopJS tests       |

---

# PR Checklist

-   [ ] commit messages adhere to the [commit message guidelines](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#what-should-the-commits-look-like)
-   [ ] review was done by a Code owner [if necessary](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#how-do-i-tell-if-my-pull-request-needs-approval-by-a-code-owner) (if you think it is not necessary, explain the reasoning in the description or in a comment)
-   [ ] `check` passes
-   [ ] `check-extended` passes
-   [ ] `rush change` [was run if applicable](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#how-do-i-describe-my-changes-for-the-changelog)

### CODE REVIEW

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/1931/ : SUCCESS in 5s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/3558/ : FAILURE in 5m 31s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/1934/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/3560/ : FAILURE in 6m 45s

**Comment by kandl** [comment]
extended test

File: libs/sdk-backend-spi/src/workspace/dateFilterConfigs/types.ts [hunk 1]
```diff
@@ -68,6 +68,20 @@ export type DateFilterGranularity =
     | "GDC.time.quarter"
     | "GDC.time.year";
 
+const dateFilterGranularity: DateFilterGranularity[] = [
+    "GDC.time.date",
```
Comment (no23reason) [inline libs/sdk-backend-spi/src/workspace/dateFilterConfigs/types.ts:72]: Why is week not there as well?

File: libs/sdk-ui-ext/src/internal/dashboardEmbedding/DashboardView/DashboardView.tsx [hunk 2]
```diff
@@ -100,6 +107,20 @@ export const DashboardView: React.FC<IDashboardViewProps> = ({
         <DashboardViewConfigProvider config={effectiveConfig}>
             <UserWorkspaceSettingsProvider settings={userWorkspaceSettings}>
                 <ColorPaletteProvider palette={colorPalette}>
+                    {isScheduledMailDialogVisible && (
+                        <ScheduledMailDialog
+                            backend={backend}
+                            workspace={workspace}
+                            locale={config?.locale}
```
Comment (no23reason) [inline libs/sdk-ui-ext/src/internal/dashboardEmbedding/DashboardView/DashboardView.tsx:114]: The locale is more involved than this, please see DashboardView component (we must default to user settings if config does not specify locale).

File: libs/sdk-ui-ext/src/internal/dashboardEmbedding/ScheduledMail/ScheduledMailDialog/RecipientsSelect/RecipientsSelect.tsx [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    /**
+     * Is enableKPIDashboardScheduleRecipients feature flag turned on?
+     */
+    enableKPIDashboardScheduleRecipients: boolean;
+
+    /**
+     * Backend to work with.
+     *
+     * Note: the backend must come either from this property or from BackendContext. If you do not specify
+     * backend here, then the executor MUST be rendered within an existing BackendContext.
```
Comment (no23reason) [inline libs/sdk-ui-ext/src/internal/dashboardEmbedding/ScheduledMail/ScheduledMailDialog/RecipientsSelect/RecipientsSelect.tsx:48]: executor is misleading here, what about component? Ditto for workspace

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/1938/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-backstop-tests/568/ : SUCCESS in 17m 25s

File: libs/sdk-ui-ext/src/internal/dashboardEmbedding/ScheduledMail/ScheduledMailDialog/ScheduledMailDialog.tsx [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+     * Note: By default, exported dashboard in the scheduled mail will use the original stored dashboard filter context,
+     * with this prop, you can override it.
+     */
+    filters?: IDashboardFilter[];
+
+    /**
+     * Backend to work with.
+     *
+     * Note: the backend must come either from this property or from BackendContext. If you do not specify
+     * backend here, then the executor MUST be rendered within an existing BackendContext.
```
Comment (no23reason) [inline libs/sdk-ui-ext/src/internal/dashboardEmbedding/ScheduledMail/ScheduledMailDialog/ScheduledMailDialog.tsx:49]: Similar to my other comments, executor is not right here.

File: libs/sdk-ui-ext/styles/internal/icons/icon-pdf.svg [hunk 5]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (no23reason) [inline libs/sdk-ui-ext/styles/internal/icons/icon-pdf.svg:1]: Should we have icons in styles? Other svgs are in src/internal/assets. I encountered this too and put my icon in `incos` folder there: https://github.com/gooddata/gooddata-ui-sdk/pull/910/files#diff-21f4416b1b7b38d4d9ebd9ea548f6fbd40e95139ea0c0f9e6db830c014526caf

File: libs/sdk-ui-ext/src/internal/dashboardEmbedding/hooks/useWorkspaceUsers.ts [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    /**
+     *  Option to filter users by the provided string.
+     */
+    search?: string;
+
+    /**
+     * Backend to work with.
+     *
+     * Note: the backend must come either from this property or from BackendContext. If you do not specify
+     * backend here, then the executor MUST be rendered within an existing BackendContext.
```
Comment (no23reason) [inline libs/sdk-ui-ext/src/internal/dashboardEmbedding/hooks/useWorkspaceUsers.ts:27]: Executor ;)

File: libs/sdk-ui-ext/src/internal/dashboardEmbedding/ScheduledMail/utils/validate.ts [hunk 7]
```diff
@@ -0,0 +1,8 @@
+// (C) 2019-2020 GoodData Corporation
+
+// eslint-disable-next-line no-useless-escape
+const EMAIL_REGEX = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
+
+export function isEmail(value: string): boolean {
```
Comment (no23reason) [inline libs/sdk-ui-ext/src/internal/dashboardEmbedding/ScheduledMail/utils/validate.ts:6]: I know it is not your code, but we should really use a library for this...

File: libs/sdk-ui-ext/src/internal/dashboardEmbedding/ScheduledMail/utils/test/scheduledMailSummary_test.ts [hunk 8]
```diff
@@ -0,0 +1,197 @@
+// (C) 2019-2020 GoodData Corporation
```
Comment (no23reason) [inline libs/sdk-ui-ext/src/internal/dashboardEmbedding/ScheduledMail/utils/test/scheduledMailSummary_test.ts:1]: Please use the naming convention of `scheduledMailSummary.test.ts` used in the rest of the repo (notice the dot). This is present in all the newly added tests in this PR.

File: libs/sdk-backend-spi/src/workspace/dateFilterConfigs/types.ts (hunk 1 above)
Comment (kandl) [reply libs/sdk-backend-spi/src/workspace/dateFilterConfigs/types.ts:72]: Good catch, thanks.

File: libs/sdk-ui-ext/src/internal/dashboardEmbedding/DashboardView/DashboardView.tsx (hunk 2 above)
Comment (kandl) [reply libs/sdk-ui-ext/src/internal/dashboardEmbedding/DashboardView/DashboardView.tsx:114]: This is actually solved on ScheduledMailDialog component level, but we should probably default it also here so it's consistent.

File: libs/sdk-ui-ext/src/internal/dashboardEmbedding/ScheduledMail/ScheduledMailDialog/RecipientsSelect/RecipientsSelect.tsx (hunk 3 above)
Comment (kandl) [reply libs/sdk-ui-ext/src/internal/dashboardEmbedding/ScheduledMail/ScheduledMailDialog/RecipientsSelect/RecipientsSelect.tsx:48]: Right, some copy paste, I'll update it.

File: libs/sdk-ui-ext/src/internal/dashboardEmbedding/ScheduledMail/utils/validate.ts (hunk 7 above)
Comment (kandl) [reply libs/sdk-ui-ext/src/internal/dashboardEmbedding/ScheduledMail/utils/validate.ts:6]: I was afraid to touch it, I think it must have some reason why it's done like this: D

File: libs/sdk-ui-ext/src/internal/dashboardEmbedding/ScheduledMail/utils/test/scheduledMailSummary_test.ts (hunk 8 above)
Comment (kandl) [reply libs/sdk-ui-ext/src/internal/dashboardEmbedding/ScheduledMail/utils/test/scheduledMailSummary_test.ts:1]: Oh, I missed utils folder :) thanks

File: libs/sdk-ui-ext/src/internal/dashboardEmbedding/hooks/useWorkspaceUsers.ts (hunk 6 above)
Comment (kandl) [reply libs/sdk-ui-ext/src/internal/dashboardEmbedding/hooks/useWorkspaceUsers.ts:27]: 🙀

File: libs/sdk-ui-ext/styles/internal/icons/icon-pdf.svg (hunk 5 above)
Comment (kandl) [reply libs/sdk-ui-ext/styles/internal/icons/icon-pdf.svg:1]: Moved

File: libs/sdk-ui-ext/src/internal/dashboardEmbedding/ScheduledMail/utils/validate.ts (hunk 7 above)
Comment (no23reason) [reply libs/sdk-ui-ext/src/internal/dashboardEmbedding/ScheduledMail/utils/validate.ts:6]: I understand :D

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Review by no23reason — APPROVED** [review]
LGTM

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-fossa-scanning-zuul-el7/625/ : FAILURE in 2m 10s (non-voting)
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-merge-compliance-check/886/ : SUCCESS in 16s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/3576/ : SUCCESS in 11m 04s

**Pull request merged**